### PR TITLE
GH-37268: [C++] adding move in some ctor in fs and dataset

### LINKED ISSUE #37268: [C++] adding move in some ctor in fs and dataset
### Describe the enhancement requested

fs and dataset sometime uses `std::shared_ptr` for copy in ctor, which is not recommended.

### Component(s)

C++

### PR DESCRIPTION
**Pull request by mapleFU**
### Rationale for this change

Fix some typo and add move in some ctor in fs and dataset.

### What changes are included in this PR?

Some minor fix.

### Are these changes tested?

No

### Are there any user-facing changes?

No

* Closes: #37268

### CODE REVIEW

**Review by raulcd — COMMENTED** [review]
Even though the change is pretty small based on our definition of MINOR this should have a GH issue as there is code change: https://github.com/apache/arrow/blob/main/CONTRIBUTING.md#minor-fixes

File: cpp/src/arrow/filesystem/filesystem.cc [hunk 1]
```diff
@@ -498,7 +498,9 @@ Result<std::string> SubTreeFileSystem::PathFromUri(const std::string& uri_string
 
 SlowFileSystem::SlowFileSystem(std::shared_ptr<FileSystem> base_fs,
                                std::shared_ptr<io::LatencyGenerator> latencies)
-    : FileSystem(base_fs->io_context()), base_fs_(base_fs), latencies_(latencies) {}
+    : FileSystem(base_fs->io_context()),
+      base_fs_(base_fs),
```
Comment (bkietz) [inline cpp/src/arrow/filesystem/filesystem.cc:502]:
```suggestion
      base_fs_(std::move(base_fs)),
```

**Review by bkietz — APPROVED** [review]
Looks good to me, thanks!

Just one more move:

**Comment by mapleFU** [comment]
I've fixed the comment, would you mind take a look again?

**Review by westonpace — APPROVED** [review]

**Pull request merged**